### ISSUE TITLE
Investigate using Auto layout for structs in the framework

### ISSUE THREAD

**Issue by GSPP** [dotnet/runtime#64261]
By default, the C# compiler marks structs `[StructLayout(LayoutKind.Sequential)]`. I believe the historic reason for this is that structs were assumed to be mainly for interop. Unfortunately, this can cause unnecessary padding (and possibly other inefficiencies?).

The default in the C# language cannot be changed because large bodies of existing code depend on this behavior.

Even if the struct itself does not benefit from `Auto`, the runtime can rearrange fields together with surrounding structs: https://github.com/dotnet/runtime/issues/7720#issuecomment-292577303. It's therefore not enough to look at a struct definition to conclude that it is not beneficial to mark it `Auto`. Profitability depends on usage as well.

I wonder if it can be considered a performance best practice to routinely mark all non-interop structs as `LayoutKind.Auto`. This strategy does not seem to have a downside except for the code clutter that this brings.

Would it be worth adopting a policy in which all eligible structs are marked as `Auto`? Should there be an analyzer for that?

I loaded all system assemblies and made lists for how the struct layout is currently set:

[5.0.8 Auto.txt](https://github.com/dotnet/runtime/files/7920392/5.0.8.Auto.txt)
[5.0.8 Explicit.txt](https://github.com/dotnet/runtime/files/7920393/5.0.8.Explicit.txt)
[5.0.8 Sequential.txt](https://github.com/dotnet/runtime/files/7920394/5.0.8.Sequential.txt)

**Comment by stephentoub** [comment]
Are there specific cases where you see there'd be a tangible benefit?  Many of the examples in your Sequential.txt are for interop and making them non-sequential would break the functionality.  Many others in Sequential.txt have only a single field and wouldn't benefit from a different layout.  Still others have all fields of the same type and there'd be no benefit to a different layout. I appreciate your interest here, but just dumping a list of all structs isn't particularly helpful.

**Comment by GSPP** [comment]
Right, I have no concrete reason to believe that any of these cases would benefit. I am not qualified to make that assessment. My thinking is more about changing policy on this. I generated the files just as a convenience for auditing the current state. It's not meant to be an actionable list.

> ... and without a lot of case law demonstrating it's worth forcing Auto to be used ...

That's probably the critical point here. Is there a systemic benefit, or is it just a small benefit here or there? You are far better able to tell than I am, so if you don't think there's anything to do here, I wouldn't mind closing the issue. I just wanted to bring this to your attention.

**Comment by AaronRobinsonMSFT** [comment]
I am dubious of the value of providing this guidance and the optimizations it may enable are likely very small or niche. That being said this does seem like an area where we should have a concrete understanding of when packing non-interop structs is beneficial. I think an investigation is warranted, whether that is done by the community or Interop team is TBD so I'll place this in "Future".

**Comment by molinch** [comment]
@stephentoub This issue brings another discussion: don't you think having an optional per project csproj configuration `<DefaultStructLayout>Auto</DefaultStructLayout>` could bring benefits while not bringing noise?

Not necessarily for the framework but developers most of the time these days don't do interop. And unsafe/pointer functionalities are really niche.

**Comment by AaronRobinsonMSFT** [comment]
> IMO, these days, 99% of the time we want Auto Layout

This is the statement that @stephentoub is alluding to when he says "Are there specific cases where you see there'd be a tangible benefit?" Why is this the case? Is there real measurable benefit? Having concrete examples and traces that show an improvement would go a long way to substantiating that statement.

**Comment by MichalPetryka** [comment]
I guess an important thing is that changing layout from Sequential (or Explicit) to Auto should be considered a binary breaking change in case some code relies on it.

**Comment by laicasaane** [comment]
In Unity, Burst compiler relies on Sequential layout and currently doesn't support Auto layout. And since Burst is going to be integrated deep into the core engine itself, this change will break it. In Unity, we need high performance really bad so unsafe is a must-have. But we don't need to utilize unsafe by ourselves since there is Burst compiler to lift the heavy work. Finally, I would like to remind you that Unity is not a niche ecosystem.

**Comment by laicasaane** [comment]
In Unity, Burst compiler relies on Sequential layout and currently doesn't support Auto layout (I can't imagine how it should support Auto). And since Burst is going to be integrated deep into the core engine itself, this change will break the whole engine. In Unity, we need high performance really bad so unsafe is a must-have. But we don't need to utilize unsafe by ourselves since there is Burst compiler to lift the heavy work. Finally, I would like to remind you that Unity is not a niche ecosystem.
